ibis/backends/clickhouse/compiler/relations.py

```diff
@@ -0,0 +1,199 @@
+import functools
+from functools import partial
+
+import sqlglot as sg
+
+import ibis.expr.datatypes as dt
+import ibis.expr.operations as ops
+from ibis.backends.clickhouse.compiler.values import translate_val
+
+
+def _needs_projection(op: ops.TableNode) -> bool:
+    return not isinstance(op, (ops.PhysicalTable, ops.Join))
+
+
+@functools.singledispatch
+def translate_rel(op: ops.TableNode, **_):
+    """Translate a table node into sqlglot."""
+    raise NotImplementedError(type(op))
+
+
+@translate_rel.register(ops.DummyTable)
+def _dummy(op: ops.DummyTable, **kw):
+    return sg.select(
+        *map(partial(translate_val, **kw), op.values), dialect="clickhouse"
+    )
+
+
+@translate_rel.register(ops.PhysicalTable)
+def _physical_table(op: ops.PhysicalTable, **_):
+    return sg.table(op.name)
+
+
+@translate_rel.register(ops.Selection)
+def _selection(op: ops.Selection, *, table, needs_alias=False, **kw):
+    # needs_alias should never be true here in explicitly, but it may get
+    # passed via a (recursive) call to translate_val
+    assert not needs_alias, "needs_alias is True"
+    if needs_alias := isinstance(op.table, ops.Join) and not isinstance(
+        op.table, (ops.LeftSemiJoin, ops.LeftAntiJoin)
+    ):
+        args = table.this.args
+        from_ = args["from"]
+        (join,) = args["joins"]
+    else:
+        from_ = join = None
+    tr_val = partial(translate_val, needs_alias=needs_alias, **kw)
+    selections = tuple(map(tr_val, op.selections)) or "*"
+    sel = sg.select(*selections, dialect="clickhouse").from_(
+        from_ if from_ is not None else table, dialect="clickhouse"
+    )
+
+    if join is not None:
+        sel = sel.join(join)
+
+    if predicates := op.predicates:
+        if join is not None:
+            sel = sg.select("*").from_(sel.subquery(kw["aliases"][op.table]))
+        res = functools.reduce(lambda x, y: f"{x} AND {y}", map(tr_val, predicates))
```

Comment:
In the `_join()` translator you are using `functools.reduce(lambda left, right: left.and_(right), preds)`, is there a difference?

Reply:
Not really. Made them consistent.